### ISSUE TITLE
perl 5.23.2: test/32-cron-next.t hangs

### ISSUE THREAD

**Issue by eserte** [niczero/mojar-cron#1]
With perl 5.23.2, the 32-cron-next.t test hangs (linux & freebsd):

```
$ perl5.23.2 -Mblib test/32-cron-next.t 
ok 1 - base time
ok 2 - every sec
ok 3 - every min
ok 4 - every 10 min
ok 5 - every 30 min 6th hour
ok 6 - midnight on 31st
ok 7 - yearly 1st Mar
(here it hangs)
```

This does not happen with other perl versions (5.23.1 and less).

**Comment by eserte** [comment]
Behavior is different when running the perl debugger: here it fails after the 7th test:

```
Not a CODE reference at /usr/perl5.23.2p/lib/5.23.2/perl5db.pl line 4236.
 at /usr/perl5.23.2p/lib/5.23.2/perl5db.pl line 4236.
        Mojar::Cron::next(Mojar::Cron=HASH(0x802fcc1b0), 1325462400) called at test/32-cron-next.t line 20
        main::next_times("0 0 * * mon", 6) called at test/32-cron-next.t line 74
```

**Comment by niczero** [comment]
Hi Slaven. 
Thank you for the report.  I'll start testing on 5.23 in response to that. 
I'll let you know once I have a plan. 
Thank you for contributing, 
Nic

-----Original Message-----

<div id="AOLMsgPart_2_fd3bd89c-4c62-451c-8f94-b9367721de20">
<div class="aolReplacedBody">
 <p>With perl 5.23.2, the 32-cron-next.t test hangs (linux & freebsd):</p> 
 <pre><code>$ perl5.23.2 -Mblib test/32-cron-next.t 
ok 1 - base time
ok 2 - every sec
ok 3 - every min
ok 4 - every 10 min
ok 5 - every 30 min 6th hour
ok 6 - midnight on 31st
ok 7 - yearly 1st Mar
(here it hangs)
</code></pre> 
 <p>This does not happen with other perl versions (5.23.1 and less).</p> 
 <p style="font-size:small;-webkit-text-size-adjust:none;color:#666;">—
Reply to this email directly or <a target="_blank" href="https://github.com/niczero/mojar-cron/issues/1">view it on GitHub</a>.<img alt="" height="1" width="1" src="https://github.com/notifications/beacon/AAY2pnqtFrjPtOXkv5eBvcjmkOcMpDcDks5ovy6LgaJpZM4F5xDs.gif"></p> 
 <div> 
  <div> 
   
 </div> 
</div>

</div>
</div>

**Comment by niczero** [comment]
Just to mention that this hasn't been fixed and makes the package uninstallable/unusable in perl 5.24.0.
I'm currently investigating.

**Comment by niczero** [comment]
Everything goes wrong at this line
`$dt->[DAY] += $slots->[0] + 7 - $dt->weekday;`
I think the problem is that 'weekday' can modify $dt -- perhaps v5.24 is optimised in a way that doesn't allow that.
An easy workaround is
`my $weekday = $dt->weekday;
$dt->[DAY] += $slots->[0] + 7 - $weekday;`

**Comment by niczero** [comment]
I think a better solution is that 'weekday' does not mutate ('normalise') its invocant.
Phew that was a tough one to get to the bottom of.
Preparing new release...

**Comment by niczero** [comment]
Fixed in v0.301
